src/utils/detached-dom-handler.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Deferred} from './promise';
+
+export class StreamHandler {
+  /**
+   *
+   */
+  constructor() {
+    const {promise, resolve} = new Deferred();
+    /** @const @private {!Promise(!Element)} */
+    this.headPromise_ = promise;
+
+    /** @const @private {!function()} */
+    this.headResolver_ = resolve;
+
+    /** @private {?Element} */
+    this.detachedBody_ = null;
+
+    /** @private {boolean} */
+    this.shouldTransfer_ = false;
+
+    /** @private {boolean} */
+    this.mergeScheduled_ = false;
+
+    // Fake vsync for glitch demo.
+    // TODO: switch to real vsync service.
+    /** @const @private */
+    this.vsync_ = {mutate: cb => Promise.resolve(cb())};
+  }
+
+  /**
+   * Callback passed into DetachedDomWriter constructor. Receives updated
+   * document everytime a new chunk is written.
+   * @param {!Document} detachedDoc
+   */
+  onChunk(detachedDoc) {
+    // <body> is newly formed.
+    if (!this.detachedBody_ && detachedDoc.body) {
+      this.detachedBody_ = detachedDoc.body;
+      this.headResolver_(detachedDoc.head);
+    }
+
+    if (this.shouldTransfer_ && !this.mergeScheduled_) {
+      this.transferBody_(detachedDoc);
+    }
+  }
+
+  /**
+   * Callback passed into DetachedDomWriter constructor. Called with complete
+   * doc when stream is closed.
+   * @param {!Document} unusedCompleteDoc
+   */
+  onEnd(unusedCompleteDoc) {
+    console.log('stream done.');
+  }
+
+  /**
+   * Promise that will resolve with <head> when available.
+   * @return {!Promise<!Element>}
+   */
+  waitForHead() {
+    return this.headPromise_;
+  }
+
+  /**
+   * Start the body transfer process. Should only be called once.
+   * Returns a promise indicating that the first body chunk has been transfered.
+   * @param {!Element} target DOM element to be appended to.
+   * @return {!Promise} resolves when first chunk has been transfered.
+   */
+  transferBody(target) {
+    if (!target) {
+      // Throw on no target given.
+    }
+
+    if (this.shouldTransfer_) {
+      // Maybe throw on subsequent calls?
+      return Promise.resolve();
+    }
+
+    this.shouldTransfer_ = true;
+    this.target_ = target;
+
+    return this.headPromise_.then(() => this.transferBody_());
+  }
+
+  /**
+   * Transfers available body elements in vsync cycle.
+   * @return {!Promise}
+   */
+  transferBody_() {
```

Comment:
So, wait, I'm confused here: is this a "transfer the whole body" or "transfer chunk of body" model? If it's "whole body", then we need to keep transferring until eof. If it's chunk - it should be named accordingly and no need for vsync - that should be the responsibility of the caller. Also, in that case, the waiting condition will be different - we'll need to wait for a next chunk. Or did you mean to use vsync.mutate as the replacement for that?

Alternative could also be to have API of `readBodyChunk():Promise<DetachedDocument>` and leave the vsync/transfer part to the caller. But these are not mutually exclusive if we get cleaner on the chunk-vs-whole thing.

Reply:
yeah should probably be `transferBodyChunk_` . My idea was that callers would just call `bodyTransfer()` expecting that this class would then take over transferring any chunks to the target when received. The `onchunk` cb would then know if `bodyTransfer` had been called and move to target immediately if so.

Reply:
Possibly some renaming and stuff needed. But otherwise, I think `vsync` should be called by caller - that'd be the main difference. Ideally we should also connect body chunks to the original stream chunks - that would reduce the need for any kind of resynchronization.

Reply:
Talked offline, the problem seemed to stem from not returning the body transfer complete promise. Refactored.